File: crawler/src/main/java/com/project/crawler/config/AwsConfig.java
package com.project.crawler.config;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import software.amazon.awssdk.auth.credentials.DefaultCredentialsProvider;
import software.amazon.awssdk.regions.Region;
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient;
import software.amazon.awssdk.services.secretsmanager.model.GetSecretValueRequest;

@Configuration
public class AwsConfig {


    @Value("${aws.secrets.name.db}")
    private String dbSecretName;

    private final ObjectMapper objectMapper;

    public AwsConfig(ObjectMapper objectMapper) {
        this.objectMapper = objectMapper;
    }

    @Bean
    public SecretsManagerClient secretsManagerClient() {
        return SecretsManagerClient.builder()
                .region(Region.AP_NORTHEAST_2)
                .credentialsProvider(DefaultCredentialsProvider.create())
                .build();
    }


    @Bean
    public String dbPassword(SecretsManagerClient client) {
        try {
            String secret = client.getSecretValue(
                    GetSecretValueRequest.builder()
                            .secretId(dbSecretName)
                            .build()
            ).secretString();

            JsonNode node = objectMapper.readTree(secret);
            return node.get("password").asText();
        } catch (Exception e) {
            throw new RuntimeException("DB 비밀번호 로드 실패", e);
        }
    }
}

File: api/src/main/kotlin/com/project/api/config/AwsConfig.kt
package com.project.api.config

import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import software.amazon.awssdk.auth.credentials.AwsCredentialsProvider
import software.amazon.awssdk.auth.credentials.AwsCredentialsProviderChain
import software.amazon.awssdk.auth.credentials.ContainerCredentialsProvider
import software.amazon.awssdk.auth.credentials.ProfileCredentialsProvider
import software.amazon.awssdk.auth.credentials.WebIdentityTokenFileCredentialsProvider
import software.amazon.awssdk.regions.Region
import software.amazon.awssdk.services.secretsmanager.SecretsManagerClient

@Configuration
class AwsConfig {

    @Bean
    fun awsCredentialsProvider(): AwsCredentialsProvider =
        AwsCredentialsProviderChain.builder()
            .addCredentialsProvider(ProfileCredentialsProvider.create())
            .addCredentialsProvider(ContainerCredentialsProvider.builder().build())
            .addCredentialsProvider(WebIdentityTokenFileCredentialsProvider.builder().build())
            .build()

    @Bean
    fun secretsManagerClient(awsCredentialsProvider: AwsCredentialsProvider): SecretsManagerClient =
        SecretsManagerClient.builder()
            .region(Region.AP_NORTHEAST_2)
            .credentialsProvider(awsCredentialsProvider)
            .build()
}
